scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerHandler.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://aws.amazon.com/apache2.0/
+ *
+ * or in the "LICENSE" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.blox.scheduling.manager;
+
+import com.amazonaws.blox.dataservicemodel.v1.client.DataService;
+import com.amazonaws.blox.dataservicemodel.v1.model.operations.ListEnvironmentsRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.operations.ListEnvironmentsResponse;
+import com.amazonaws.blox.lambda.LambdaFunction;
+import com.amazonaws.blox.scheduling.scheduler.SchedulerInput;
+import com.amazonaws.blox.scheduling.scheduler.SchedulerOutput;
+import com.amazonaws.blox.scheduling.state.ClusterSnapshot;
+import com.amazonaws.blox.scheduling.state.ECSState;
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+import com.spotify.futures.CompletableFutures;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__({@Autowired}))
+@Log4j2
+public class ManagerHandler implements RequestHandler<ManagerInput, ManagerOutput> {
+
+  final DataService data;
+  final ECSState ecs;
+  final LambdaFunction<SchedulerInput, SchedulerOutput> scheduler;
+
+  @Override
+  public ManagerOutput handleRequest(ManagerInput input, Context context) {
+    ListEnvironmentsResponse r =
+        data.listEnvironments(new ListEnvironmentsRequest(input.getClusterArn()));
+    List<String> environments = r.getEnvironmentArns();
+
+    ClusterSnapshot state = ecs.snapshotState(input.getClusterArn());
+
+    List<SchedulerOutput> schedulerResults =
+        environments
+            .stream()
+            .map(environmentId -> scheduler.callAsync(new SchedulerInput(state, environmentId)))
+            .collect(CompletableFutures.joinList())
```

Comment:
so how does this work? does this block until all the results are available?

Reply:
It will rapidly call `callAsync` for all environments, which immediately returns a `CompletableFuture`. The `.collect(CompletableFutures.joinList())` line will then efficiently wait for all the futures to complete, and return a single `CompletableFuture` that yields the list of all the values of the futures from `callAsync`.

Thinking about it, I think it's worth breaking this up to make it clearer what's happening like so:

```java
Stream<CompletableFuture<SchedulerOutput>> pendingRequests = environments.stream().map(...);
CompletableFuture<List<SchedulerOutput>> pendingResponses = pendingRequests.collect(CompletableFutures.joinList());
List<SchedulerOutput> results = pendingResponses.join();
```

Would that be clearer?

Reply:
So it does wait for futures to complete and return a result? I don't think that's unclear what I'm asking is then this is essentially a blocking method, correct?

Reply:
Yes, correct, this method will block on the last `.join()` of the composite future until all requests are complete.